wallet/db.c

```diff
@@ -223,27 +223,37 @@ static void close_db(struct db *db) { sqlite3_close(db->sql); }
 
 bool db_begin_transaction(struct db *db)
 {
-	assert(!db->in_transaction);
-	/* Clear any errors from previous transactions and
-	 * non-transactional queries */
-	db_clear_error(db);
-	db->in_transaction = db_exec(__func__, db, "BEGIN TRANSACTION;");
+	if (!db->in_transaction) {
+		/* Clear any errors from previous transactions and
+		 * non-transactional queries */
+		db_clear_error(db);
+		db->in_transaction = db_exec(__func__, db, "BEGIN TRANSACTION;");
+		db->transaction_depth = 1;
+	} else {
+		db->transaction_depth++;
+	}
 	return db->in_transaction;
 }
 
```

Comment:
Now you can't rollback a transaction safely, because it might be nested :(  Our db model assumes no spurious commit failures (we don't loop), so you'll just break the outer transaction if you try.

Callers *DO* need to track what's happening.  This patch is a classic case of making it easier to use, but also making it easier to misuse :(